### PR TITLE
Cross margin accounts and other futures / state related refactor

### PR DESCRIPTION
**Pull request by avclarke**
**Refactor futures state logic**
This moves the remaining futures recoil state over to redux along with migrating the related queries to the sdk.

**Account separation**
Cross margin account specific data is now nested by account and network, this means as users switch networks and wallets the display data will change immediately to match the network / account context.

**Implement Redux persist**
Redux persist allows us to persist state between sessions by storing in local storage and hydrating state when the app is loaded.

_State migrations - Now that state is persisted it's important when any existing state changes are made a [migration](https://github.com/rt2zz/redux-persist#migrations) is included to avoid errors occurring._ 

**Implement Error Notifications**
There is now a function called `notifyError` which takes a message as parameter and will display an error notification to the user. This uses a similar toast component to our transaction components. This relates to issues #1298 #1453 but there needs to be further work to fully implement the error notifications throughout the app.

closes: #1856

### CODE REVIEW

File: sections/futures/TradeCrossMargin/EditCrossMarginLeverageModal.tsx [hunk 1]
```diff
@@ -49,9 +48,8 @@ type DepositMarginModalProps = {
 
 export default function EditLeverageModal({ onDismiss, editMode }: DepositMarginModalProps) {
```
Comment (Tburm) [inline sections/futures/TradeCrossMargin/EditCrossMarginLeverageModal.tsx:49]: Having trouble with this component when I don't have a position open.

With no position, I click "edit" to open the modal. When I click confirm, it throws an error using the notifier. I can't click on screen to dismiss the modal, and clicking the X button doesn't dismiss the modal.

File: sections/futures/TradeCrossMargin/EditCrossMarginLeverageModal.tsx [hunk 2]
```diff
@@ -122,56 +119,8 @@ export default function EditLeverageModal({ onDismiss, editMode }: DepositMargin
 	);
 
 	const onConfirm = useCallback(async () => {
-		setError(null);
-		if (editMode === 'existing_position' && position?.position) {
```
Comment (Tburm) [inline sections/futures/TradeCrossMargin/EditCrossMarginLeverageModal.tsx:126]: losing this logic means it tries to submit a transaction when in "new position" mode.

File: sections/futures/TradeCrossMargin/EditCrossMarginLeverageModal.tsx [hunk 3]
```diff
@@ -122,56 +119,8 @@ export default function EditLeverageModal({ onDismiss, editMode }: DepositMargin
...
-						},
-					});
-				}
-			} catch (err) {
-				setSubmitting(false);
-				setError(t('common.transaction.transaction-failed'));
-				logError(err);
-			}
-			resetTradeState();
-		} else {
```
Comment (Tburm) [inline sections/futures/TradeCrossMargin/EditCrossMarginLeverageModal.tsx:155]: I think we need to preserve this logic

File: sections/futures/PositionCard/PositionCard.tsx [hunk 4]
```diff
@@ -89,10 +88,8 @@ const PositionCard: React.FC<PositionCardProps> = () => {
 			: undefined;
 
 	const thisPositionHistory = useMemo(() => {
-		return positionHistory[futuresAccountType].find(
-			({ asset, isOpen }) => isOpen && asset === marketAsset
-		);
-	}, [positionHistory, marketAsset, futuresAccountType]);
+		return positionHistory.find(({ asset, isOpen }) => isOpen && asset === marketAsset);
+	}, [positionHistory, marketAsset]);
```
Comment (Tburm) [inline sections/futures/PositionCard/PositionCard.tsx:92]: We reuse this logic in a few different components, maybe should just have a `currentPositionHistory` selector that takes care of this logic in state.

File: sections/futures/Trade/ManagePosition.tsx [hunk 5]
```diff
@@ -8,15 +8,14 @@ import Loader from 'components/Loader';
 import { useFuturesContext } from 'contexts/FuturesContext';
 import { previewErrorI18n } from 'queries/futures/constants';
 import { setOpenModal } from 'state/app/reducer';
-import { selectOpenModal } from 'state/app/selectors';
```
Comment (Tburm) [inline sections/futures/Trade/ManagePosition.tsx:11]: Comment is for this component. We display the `orderError` here, should we still display transaction errors even though they would display in the notifier:
image

Not related to changes you made, but an improvement to displaying the unformatted errors.

File: sections/futures/TradeCrossMargin/EditCrossMarginLeverageModal.tsx (hunk 1 above)
Comment (Tburm) [reply sections/futures/TradeCrossMargin/EditCrossMarginLeverageModal.tsx:49]: Specifically when opening this way:
image

When opened from an open position it works as expected.

File: queries/futures/types.ts [hunk 6]
```diff
@@ -25,37 +25,6 @@ export type MarginTransfer = {
...
-	side: PositionSide;
-	pnl: Wei;
-	pnlWithFeesPaid: Wei;
-	totalVolume: Wei;
-	trades: number;
-};
-
 export enum PositionSide {
 	LONG = 'long',
 	SHORT = 'short',
```
Comment (Tburm) [inline queries/futures/types.ts:30]: Looked like we were still importing this one in some places, might be worth removing this as well and always importing from SDK.

File: sections/futures/PositionCard/PositionCard.tsx (hunk 4 above)
Comment (avclarke) [reply sections/futures/PositionCard/PositionCard.tsx:92]: yea good call

File: sections/futures/Trade/ManagePosition.tsx (hunk 5 above)
Comment (avclarke) [reply sections/futures/Trade/ManagePosition.tsx:11]: Yea makes sense, I've removed the transaction error for now but eventually when we do some more clean up around errors we probably want to push all these order / preview errors through the notifier.

File: state/store.ts [hunk 7]
```diff
@@ -46,4 +73,8 @@ export type AppThunk<ReturnType = void> = ThunkAction<
 	AnyAction
 >;
 
+export const persistor = persistStore(store);
+
+// persistor.purge();
```
Comment (koredefashokun) [inline state/store.ts:78]: I think you forgot this comment.

File: state/types.ts [hunk 8]
```diff
@@ -11,7 +11,7 @@ export enum FetchStatus {
 
 export type QueryStatus = {
 	status: FetchStatus;
-	error: string | null | undefined;
+	error?: string | null | undefined;
```
Comment (koredefashokun) [inline state/types.ts:14]: We can omit the explicit `undefined` since this is optional.

**Review by koredefashokun — COMMENTED** [review]
Changes look very good! I left a few comments around small things, but I'm pretty excited about the changes (especially about getting rid of the `RefetchContext`).

File: sdk/utils/futures.ts [hunk 9]
```diff
@@ -86,7 +84,7 @@ export const marketsForNetwork = (networkId: number) => {
 		case 420:
 			return TESTNET_MARKETS;
 		default:
-			logError('Futures is not supported on this network.');
+			logError(new Error('Futures is not supported on this network.'));
```
[the remark this answers is not in the document]
Comment (koredefashokun) [reply sdk/utils/futures.ts:87]: By the way, I noticed why this is this way, so you can ignore.

File: sdk/queries/futures.ts [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+
+import { getFuturesEndpoint } from 'sdk/utils/futures';
+
+export const queryAccountsFromSubgraph = async (
+	networkId: NetworkId,
+	walletAddress: string | null
+): Promise<string[]> => {
+	if (!walletAddress) return [];
+
+	const futuresEndpoint = getFuturesEndpoint(networkId);
```
[the remark this answers is not in the document]
Comment (avclarke) [reply sdk/queries/futures.ts:13]: Yea good call, there's already a getter on the futures sdk service `futuresGqlEndpoint` so should probably just use that instead.

File: components/Error/ErrorNotifier.tsx [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+	if (!message) return '';
+	if (message.includes('insufficient funds for intrinsic transaction cost'))
+		return 'Insufficient ETH balance for gas cost';
+	return message;
+};
+
+const StyledToastContainer = styled(ToastContainer)`
+	.Toastify__toast-container {
+		background-color: ${(props) => props.theme.colors.navy};
+		border: 1px solid ${(props) => props.theme.colors.green};
```
[the remark this answers is not in the document]
Comment (avclarke) [reply components/Error/ErrorNotifier.tsx:45]: Seems these styles aren't actually relevant, will just remove, good spot though was some copy paste

File: sdk/queries/futures.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (avclarke) [reply sdk/queries/futures.ts:13]: Thinking about this some more I'm going to leave for now as the next step will be to move all these queries over to the sdk anyway.

**Review by platschi — APPROVED** [review]

**Review by Tburm — COMMENTED** [review]
All looks good except one issue. I opened this modal to adjust leverage on an open cross margin position:
image

It doesn't look like there is a way to close the modal. The X doesn't work, clicking off the modal doesn't work.

**Review by platschi — APPROVED** [review]

**Pull request merged**